### PR TITLE
Design Doc for Regularization

### PR DESCRIPTION
**Pull request by abhinavarora**

### CODE REVIEW

File: doc/design/regularization.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    - Makes it easy for the users to customize and extend the framework. 
+
+	**Disadvantages**:
+    - Implementation requires comprehensive design and time. 
+
+## Proposal for Regularization in PaddlePaddle
+
+### Low-Level implementation
+
+In the new design, we propose to create new operations for regularization. For now, we can add 2 ops thgat correspond to the most frequently used regularizations:
```
Comment (jacquesqiao) [inline doc/design/regularization.md]: thgat   ==>  that

File: doc/design/regularization.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+##### L1 Regularization
+<img src="./images/l1_regularization.png" align="center"/><br/>
+
+A much more detailed mathematical background of reguilarization can be found [here](http://www.deeplearningbook.org/contents/regularization.html).
+
+
+## How to do Regularization in PaddlePaddle
+
+On surveying existing frameworks like Tensorflow, PyTorch, Caffe, etc, it can be seen that there are 2 common approaches of doing regularization:
```
Comment (dzhwinter) [inline doc/design/regularization.md]: Thanks for the comparison about torch, tensorflow!

Maybe the survey part can be put into an issue or wiki page and add a reference to the bottom. e.g [dependency engine](https://github.com/PaddlePaddle/Paddle/wiki/mxnet::engine-survey). 

My point is the design doc should be simplified enough for the users who want to learn the design detail. If they want more details about the design decision, they can find the related survey/discussion easily. But we should not put them into design doc.

File: doc/design/regularization.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
+# Regularization in PaddlePaddle
+
+## Introduction to Regularization
+A central problem in machine learning is how to design an algorithm that will perform well not just on the training data, but also on new data. Many strategies are used by machine learning practitioners to reduce the test error, possibly at the expense of increased training error. These strategies are collectively known as **regularization**. 
+
```
Comment (dzhwinter) [inline doc/design/regularization.md]: Maybe we can introduce the `overfit` problem and then comes out the `regularization`.
>Many strategies are used by machine learning practitioners to reduce the test error, possibly at the expense of increased training error.

These strategies summarization of `regularization` is not very accurate to my first glance.  
from the wikipedia, 
[regularization](https://en.wikipedia.org/wiki/Regularization_(mathematics))

>In general, regularization is a technique that applies to objective functions in ill-posed problems formulated as optimization problems

File: doc/design/regularization.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+The proposal is to add these ops in a lazy manner just before the backward pass. 
+
+#### Storage of Regularization attributes
+
+Since we want to create the regularization ops in a lazy manner, the regularization attributes (type of regularization and weight of regularization penalty) can be stored as attributes of the [`Parameter`](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/framework/framework.py#L421) class. This is because regularization is a property of the parameters and storing regularization properties with Parameters also allows for shared parameters. 
+
+#### High-level API
+
+In PaddlePaddle Python API, users will primarily rely on [layer functions](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/python_api.md#layer-function) to create neural network layers. Hence, we lso need to provide regularization functionality in layer functions. The design of these APIs can be postponed for later right now. A good reference for these APIs can be found in [Keras](https://keras.io/regularizers/) and also by looking at Tensorflow in [`tf.contrib.layers`](https://www.tensorflow.org/api_guides/python/contrib.layers).
```
Comment (dzhwinter) [inline doc/design/regularization.md]: Not related this PR, I just want to remind that our `V2 API` is a determined interface. We must compatible with it. I'm not sure the layer function will take this job or regularization function also need to consider it.

File: doc/design/regularization.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+<img src="./images/feed_forward_regularized.png" align="center"/><br/>
+   
+### Python API implementation for Regularization
+
+Using the low level ops, `L2_regularization_op` and `L1_regularization_op`, any user can add regularization to their computation graphs. However, this will require a lot of lines of code and we should design Python APIs that support regularization. An example of such an API can be seen in [Keras](https://keras.io/regularizers/). As per the PaddlePaddle [Python API design](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/python_api.md), the layer functions are responsible for creating operators, operator parameters and variables. Since regularization is a property of parameters, it makes sense to create these in the layer functions. 
+
+#### Creation of Regularization ops
+There are two possibilities for creating the regularization ops:
+1. We create these ops immediately while building the computation graph.
```
Comment (dzhwinter) [inline doc/design/regularization.md]: Maybe we should write down the proposal one. The same reason with above comment, we can put these two options in an issue or somewhere, then we discuss or vote for a better choice.

And for here it will definitely be option 2. 

Because when we run a specific target, some operators should not be included, for example, the googlenet(inception model) run different level optimized targets, or when do the serving job such as inference examples.

File: doc/design/regularization.md (hunk 4 above)
Comment (abhinavarora) [reply doc/design/regularization.md]: Hi @dzhwinter We might have to discuss this offline because even I had this concern. However, I saw that we are also changing the interface for the python Optimizers.

Comment (lcy-seso) [reply doc/design/regularization.md]: I just see this PR. As I understand, the regularizers only have gradient operators. It is no need to create the forward operator which causes computations playing not important roles. The l2 regularizer can be easily implemented in the optimizer, but L1 is very special.

Comment (abhinavarora) [reply doc/design/regularization.md]: @lcy-seso  I agree with you that the L2 regularizer can be easily implemented in the optimizer. However, I believe that implementing in the optimizer can be an added feature that Paddle could support. Having regularization ops can be a more general way for us and Paddle users to implement custom regularization schemes. Let me clean this doc for typos and then we can vote/discuss on both the options?

Comment (lcy-seso) [reply doc/design/regularization.md]: No problem. I am also curious about and interested in how we are going to implement the L1 regularization.
